### PR TITLE
page styling

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
allow page styling

`with [tgb.Page](http://tgb.page/)() as page`
 or 
`page = Markdown(""" ... """)`
 or 
`page = HTML(....)`

```
page.set_style(
    {
        ".taipy-chat-sent .MuiPaper-root": {"color": "var(--color-contrast)", "padding-top": "unset"},
        ".taipy-chat-sent .MuiBox-root": {"color": "var(--color-contrast)"},
    }
)
```
and support nested css
```
page.set_style(
{
        ".taipy-chat-sent": {
            ".MuiPaper-root": {"color": "var(--color-contrast)", "padding-top": "unset"},
            ".MuiBox-root": {"color": "var(--color-contrast)"},
        },
    }
)
```

### CODE REVIEW

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

File: taipy/gui/page.py [hunk 1]
```diff
@@ -140,3 +141,10 @@ def render(self, gui) -> str:
         if self._renderer is not None:
             return self._renderer.render(gui)
         return "<h1>No renderer found for page</h1>"
+
+    def set_style(self, style: t.Dict[str, t.Dict[str, str]]):
+        self.__style = style if isinstance(style, dict) else None
+        return self
+
+    def get_style(self):
```
Comment (FabienLelaquais) [inline taipy/gui/page.py]: What would you expose ' get_style()'  for?

File: taipy/gui/page.py [hunk 2]
```diff
@@ -140,3 +141,10 @@ def render(self, gui) -> str:
         if self._renderer is not None:
             return self._renderer.render(gui)
         return "<h1>No renderer found for page</h1>"
+
+    def set_style(self, style: t.Dict[str, t.Dict[str, str]]):
```
Comment (FabienLelaquais) [inline taipy/gui/page.py]: Doc missing?

File: taipy/gui/server.py [hunk 3]
```diff
@@ -229,6 +230,7 @@ def _render(self, html_fragment, style, head, context):
         template_str = _Server.__RE_CLOSING_CURLY.sub(_Server.__CLOSING_CURLY, template_str)
         template_str = template_str.replace('"{!', "{")
         template_str = template_str.replace('!}"', "}")
+        style = get_style(style)
```
Comment (FabienLelaquais) [inline taipy/gui/server.py:233]: -> `_get_style()`?

File: taipy/gui/utils/_css.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+
+import os
+import typing as t
+
+
+def get_style(style: t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, str]]]]):
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_css.py]: I'm afraid it's worse than that...
t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, t.Union[str, t.Dict[str, str]]]]]]]] why not?

File: taipy/gui/utils/_css.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    if isinstance(style, dict):
+            style_arr = []
+            for k, v in style.items():
+                if isinstance(v, dict):
+                    rules= []
+                    for vk, vv in v.items():
+                        if isinstance(vv, dict):
+                             rules.append(get_style({vk: vv}))
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_css.py:14]: 👍

File: taipy/gui/utils/_css.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+                        else:
+                            rules.append(f'{vk}:{vv};')
+                    if rules:
+                        style_arr.append(f"{k}{{{''.join(rules)}}}")
+            return os.linesep.join(style_arr) if style_arr else None
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_css.py:19]: Do you really need a ' real'  line skip?

**Comment by FredLL-Avaiga** [comment]
All based on your work @dinhlongviolin1

**Review by FabienLelaquais — DISMISSED** [review]
I love it!

File: taipy/gui/page.py (hunk 1 above)
Comment (dinhlongviolin1) [reply taipy/gui/page.py]: for the Gui class to call it. still technically a public method for other class to call it. wouldnt mind to make it private though

File: taipy/gui/utils/_css.py (hunk 4 above)
Comment (dinhlongviolin1) [reply taipy/gui/utils/_css.py]: imagine writing that in java

File: taipy/gui/utils/_css.py (hunk 6 above)
Comment (dinhlongviolin1) [reply taipy/gui/utils/_css.py:19]: should be easier to debug on the user side, they can go to the console and look it up

File: taipy/gui/page.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/gui/page.py]: 👍

File: taipy/gui/utils/_css.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_css.py]: no need to be too precise (we can't anyway)

File: taipy/gui/utils/_css.py (hunk 6 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_css.py:19]: just to make things a bit more readable indeed

**Review by namnguyen20999 — DISMISSED** [review]

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

**Review by dinhlongviolin1 — APPROVED** [review]
Works for me

**Review by namnguyen20999 — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**